### PR TITLE
[Backport] Refine Uncommitter and Introduce ZUnmapper

### PR DESCRIPTION
**Pull request by tanghaoth90**

### CODE REVIEW

File: src/hotspot/os/linux/gc/z/zPhysicalMemoryBacking_linux.cpp [hunk 1]
```diff
@@ -304,18 +304,18 @@ void ZPhysicalMemoryBacking::warn_available_space(size_t max) const {
   // Warn if the filesystem doesn't currently have enough space available to hold
   // the max heap size. The max heap size will be capped if we later hit this limit
   // when trying to expand the heap.
-  if (_available < max) {
+  if (_available < max_capacity) {
     log_warning(gc)("***** WARNING! INCORRECT SYSTEM CONFIGURATION DETECTED! *****");
     log_warning(gc)("Not enough space available on the backing filesystem to hold the current max Java heap");
     log_warning(gc)("size (" SIZE_FORMAT "M). Please adjust the size of the backing filesystem accordingly "
-                    "(available", max / M);
+                      "(available", max_capacity / M);
```
Comment (linade) [inline src/hotspot/os/linux/gc/z/zPhysicalMemoryBacking_linux.cpp:311]: indent wrong

File: src/hotspot/os/linux/gc/z/zPhysicalMemoryBacking_linux.cpp [hunk 2]
```diff
@@ -304,18 +304,18 @@ void ZPhysicalMemoryBacking::warn_available_space(size_t max) const {
...
     log_warning(gc)("space is currently " SIZE_FORMAT "M). Continuing execution with the current filesystem "
-                    "size could", _available / M);
-    log_warning(gc)("lead to a premature OutOfMemoryError being thrown, due to failure to map memory.");
+                      "size could", _available / M);
```
Comment (linade) [inline src/hotspot/os/linux/gc/z/zPhysicalMemoryBacking_linux.cpp:313]: indent wrong

File: src/hotspot/os/linux/gc/z/zPhysicalMemoryBacking_linux.cpp [hunk 3]
```diff
@@ -338,24 +338,24 @@ void ZPhysicalMemoryBacking::warn_max_map_count(size_t max) const {
...
   // In the worst cases, ZGC will map each granule three times, i.e. once per heap view.
   // We speculate that we need another 20% to allow for non-ZGC subsystems to map memory.
-  const size_t required_max_map_count = (max / ZGranuleSize) * 3 * 1.2;
+  const size_t required_max_map_count = (max_capacity / ZGranuleSize) * 3 * 1.2;
   if (actual_max_map_count < required_max_map_count) {
     log_warning(gc)("***** WARNING! INCORRECT SYSTEM CONFIGURATION DETECTED! *****");
     log_warning(gc)("The system limit on number of memory mappings per process might be too low for the given");
     log_warning(gc)("max Java heap size (" SIZE_FORMAT "M). Please adjust %s to allow for at",
-                    max / M, filename);
+                      max_capacity / M, filename);
```
Comment (linade) [inline src/hotspot/os/linux/gc/z/zPhysicalMemoryBacking_linux.cpp:346]: indent

File: src/hotspot/os/linux/gc/z/zPhysicalMemoryBacking_linux.cpp [hunk 4]
```diff
@@ -338,24 +338,24 @@ void ZPhysicalMemoryBacking::warn_max_map_count(size_t max) const {
...
     log_warning(gc)("least " SIZE_FORMAT " mappings (current limit is " SIZE_FORMAT "). Continuing execution "
-                    "with the current", required_max_map_count, actual_max_map_count);
-    log_warning(gc)("limit could lead to a fatal error, due to failure to map memory.");
+                      "with the current", required_max_map_count, actual_max_map_count);
```
Comment (linade) [inline src/hotspot/os/linux/gc/z/zPhysicalMemoryBacking_linux.cpp:348]: indent

File: src/hotspot/share/gc/z/zArguments.cpp [hunk 5]
```diff
@@ -66,6 +66,20 @@ void ZArguments::initialize() {
     vm_exit_during_initialization("The flag -XX:+UseZGC can not be combined with -XX:ConcGCThreads=0");
   }
 
+  // Select medium page size so that we can calculate the max reserve
+  ZHeuristics::set_medium_page_size();
+
+/*
```
Comment (linade) [inline src/hotspot/share/gc/z/zArguments.cpp:72]: Do we need to keep the commented code?

**Review by linade — APPROVED** [review]

**Review by mmyxym — APPROVED** [review]

**Pull request merged**